### DISCUSSION TITLE
Anomalies in the KB?

### DISCUSSION THREAD

**Discussion by VitaliyMez** [denkbares/d3web-KnowWE#14] [Ideas]
Hi, please advise any tool to find anomalies in the KB, eg. to debug contradictory rules.
Thank you!

**Comment by AlbrechtStriffler** [comment]
Did you have a look at the CIDashboard-Markup? https://www.d3web.de/Wiki.jsp?page=Doc%20CIDashboard
If I'm not mistaken, there isn't yet a test for contradictory rules... maybe if you tell me more about the specific issue you have...

**Reply by VitaliyMez** [reply]
Dear Albrecht,

As far as I see it in CIDashboard you have an Abnormality test for DiaFlux, I think it may be the closest to my problem...

I have a set of IF-THEN-ELSE rules, setting values to variables. 
Sometimes we may have a contradictory assignment, due to firing two or more interconnecting rules. 
How can I track it?

Thank you!

**Comment by AlbrechtStriffler** [comment]
Hi Vitaliy,

we currently don't have CI tests checking contradictory assignments, because this usually isn't an actual problem in our use cases and, depending on the conditions of the rules, is also hard to track in a helpful way. Also, for some question types, it is normal to have multiple assignments at the same time, because values and facts can and should be merged. For example, multiple assignments to the same [num] question will sum up the values of the assignmet, for [mc] questions, the choices will be aggregated...

THEN questionNum = 5 
THEN questionNum = 2
at the same time will result in questionNum = 7

THEN questionMC = choice1
THEN questionMC = choice2
at the same time will result in questionMC = choice1, choice2

The same is true of course for diagnosis rules using Solutions, also see https://www.d3web.de/Wiki.jsp?page=Doc%20DiagnosisRule

What you can do right now to track assignments is to use the "Show Info"-Tool on the questions, there you will see all rules using the same question or solution.
![image](https://user-images.githubusercontent.com/13681360/201180033-c0a2c77c-bba3-40f4-8018-abc549327eb5.png)
